Panic error during reconciliation of awsmachine

/kind bug

1. Create the management cluster
2. Create Cluster API infrastructure
3. Create CABPK infrastructure
4. Create Cloud user stack
5. Set Target Cloud Credentials as Environment variables
6. Create Cloud(eg.AWS) infrastructure
7. Create Cluster
8. Create control plane machine
9. Delete the EC2 instance which is referenced in worker node, manually.

`I1220 09:35:53.883037       1 awsmachine_controller.go:261] controllers/AWSMachine "msg"="Reconciling AWSMachine" "awsCluster"="test-infra-814zqm" "awsMachine"="test-infra-1ncyj8-0" "cluster"="test-2ssj10" "machine"="test-k9f71y-0" "namespace"="test-uc28bd" 
E1220 09:35:55.337500       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 268 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1879e00, 0x2e31850)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:51 +0x82
panic(0x1879e00, 0x2e31850)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/ec2.(*Service).SDKToInstance(0xc000124218, 0xc000256c00, 0xc00051c300, 0x0, 0x0)
	/workspace/pkg/cloud/services/ec2/instances.go:581 +0x4ac
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/ec2.(*Service).InstanceIfExists(0xc000124218, 0xc0003a8ef0, 0xc0003c3fb0, 0x0, 0x0)
	/workspace/pkg/cloud/services/ec2/instances.go:96 +0x268
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).findInstance(0xc000317fb0, 0xc00048ef00, 0xc000124218, 0xc000595768, 0x40c308, 0x8)
	/workspace/controllers/awsmachine_controller.go:244 +0x190
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).getOrCreate(0xc000317fb0, 0xc00048ef00, 0xc000124218, 0x1b06fcc, 0x2a, 0x1)
	/workspace/controllers/awsmachine_controller.go:370 +0x43
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).reconcileNormal(0xc000317fb0, 0x1e08a80, 0xc0000b0068, 0xc00048ef00, 0xc000230cb0, 0x0, 0x1e27520, 0xc000270f00, 0x1e19d20)
	/workspace/controllers/awsmachine_controller.go:288 +0x1c3
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).Reconcile(0xc000317fb0, 0xc000333f00, 0xb, 0xc00063a320, 0x1e, 0x2e4a500, 0x0, 0x0, 0x0)
	/workspace/controllers/awsmachine_controller.go:144 +0xa62
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00027e140, 0x18fda00, 0xc0005260e0, 0x18fda00)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:216 +0x146
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00027e140, 0xc0002a4700)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:192 +0xb5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00027e140)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:171 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000121d00)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000121d00, 0x3b9aca00, 0x0, 0xc00011a901, 0xc00009a5a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000121d00, 0x3b9aca00, 0xc00009a5a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.3.0/pkg/internal/controller/controller.go:157 +0x311
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x14a20fc]

`

Reconcile process should detect EC2 instance termination and corresponding awsmachine & machine objects should be updated with error message. And also machine object phase should be `failed`

This was found while running E2E automation on `release-0.4` branch.

**Environment:**

- Cluster-api-provider-aws version: release-0.4
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):  darwin amd64
[capa-log.txt](https://github.com/kubernetes-sigs/cluster-api-provider-aws/files/3987696/capa-log.txt)

--- a/pkg/cloud/services/ec2/instances.go
+++ b/pkg/cloud/services/ec2/instances.go
@@ -560,7 +560,7 @@ func (s *Service) getInstanceRootDeviceSize(instance *ec2.Instance) (*int64, err
 			return out.Volumes[0].Size, nil
 		}
 	}
-	return nil, nil
+	return nil, errors.Errorf("no root volume found for EC2 instance %q", *instance.InstanceId)
 }
 
 // SDKToInstance converts an AWS EC2 SDK instance to the CAPA instance type.
@@ -603,7 +603,6 @@ func (s *Service) SDKToInstance(v *ec2.Instance) (*infrav1.Instance, error) {
 	if err != nil {
 		return nil, errors.Wrapf(err, "unable to get root volume size for instance: %q", aws.StringValue(v.InstanceId))
 	}
-
 	i.RootDeviceSize = aws.Int64Value(rootSize)
 
 	i.Addresses = s.getInstanceAddresses(v)
@@ -613,7 +612,6 @@ func (s *Service) SDKToInstance(v *ec2.Instance) (*infrav1.Instance, error) {
 
 func (s *Service) getInstanceAddresses(instance *ec2.Instance) []corev1.NodeAddress {
 	addresses := []corev1.NodeAddress{}
-
 	for _, eni := range instance.NetworkInterfaces {
 		privateDNSAddress := corev1.NodeAddress{
 			Type:    corev1.NodeInternalDNS,

--- a/pkg/cloud/services/ec2/instances_test.go
+++ b/pkg/cloud/services/ec2/instances_test.go
@@ -87,11 +87,31 @@ func TestInstanceIfExists(t *testing.T) {
 											Code: aws.Int64(16),
 											Name: aws.String(ec2.StateAvailable),
 										},
+										RootDeviceName: aws.String("device-1"),
+										BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+											{
+												DeviceName: aws.String("device-1"),
+												Ebs: &ec2.EbsInstanceBlockDevice{
+													VolumeId: aws.String("volume-1"),
+												},
+											},
+										},
 									},
 								},
 							},
 						},
 					}, nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {
@@ -115,6 +135,42 @@ func TestInstanceIfExists(t *testing.T) {
 					InstanceIds: []*string{aws.String("one")},
 				}).
 					Return(nil, errors.New("some unknown error"))
+			},
+			check: func(i *infrav1.Instance, err error) {
+				if err == nil {
+					t.Fatalf("expected an error but got none.")
+				}
+			},
+		},
+		{
+			name:       "instance without root volume, returns error",
+			instanceID: "one",
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeInstances(&ec2.DescribeInstancesInput{
+					InstanceIds: []*string{aws.String("one")},
+				}).
+					Return(&ec2.DescribeInstancesOutput{
+						Reservations: []*ec2.Reservation{
+							{
+								Instances: []*ec2.Instance{
+									{
+										InstanceId:   aws.String("id-1"),
+										InstanceType: aws.String("m5.large"),
+										SubnetId:     aws.String("subnet-1"),
+										ImageId:      aws.String("ami-1"),
+										IamInstanceProfile: &ec2.IamInstanceProfile{
+											Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
+										},
+										State: &ec2.InstanceState{
+											Code: aws.Int64(16),
+											Name: aws.String(ec2.StateAvailable),
+										},
+										RootDeviceName: aws.String("404-not-found"),
+									},
+								},
+							},
+						},
+					}, nil)
 			},
 			check: func(i *infrav1.Instance, err error) {
 				if err == nil {
@@ -319,15 +375,35 @@ func TestCreateInstance(t *testing.T) {
 								IamInstanceProfile: &ec2.IamInstanceProfile{
 									Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
 								},
-								InstanceId:   aws.String("two"),
-								InstanceType: aws.String("m5.large"),
-								SubnetId:     aws.String("subnet-1"),
-								ImageId:      aws.String("ami-1"),
+								InstanceId:     aws.String("two"),
+								InstanceType:   aws.String("m5.large"),
+								SubnetId:       aws.String("subnet-1"),
+								ImageId:        aws.String("ami-1"),
+								RootDeviceName: aws.String("device-1"),
+								BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+									{
+										DeviceName: aws.String("device-1"),
+										Ebs: &ec2.EbsInstanceBlockDevice{
+											VolumeId: aws.String("volume-1"),
+										},
+									},
+								},
 							},
 						},
 					}, nil)
 				m.WaitUntilInstanceRunningWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
 					Return(nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {
@@ -422,16 +498,36 @@ func TestCreateInstance(t *testing.T) {
 								IamInstanceProfile: &ec2.IamInstanceProfile{
 									Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
 								},
-								InstanceId:   aws.String("two"),
-								InstanceType: aws.String("m5.large"),
-								SubnetId:     aws.String("subnet-3"),
-								ImageId:      aws.String("ami-1"),
+								InstanceId:     aws.String("two"),
+								InstanceType:   aws.String("m5.large"),
+								SubnetId:       aws.String("subnet-3"),
+								ImageId:        aws.String("ami-1"),
+								RootDeviceName: aws.String("device-1"),
+								BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+									{
+										DeviceName: aws.String("device-1"),
+										Ebs: &ec2.EbsInstanceBlockDevice{
+											VolumeId: aws.String("volume-1"),
+										},
+									},
+								},
 							},
 						},
 					}, nil)
 
 				m.WaitUntilInstanceRunningWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
 					Return(nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {
@@ -539,15 +635,36 @@ func TestCreateInstance(t *testing.T) {
 								IamInstanceProfile: &ec2.IamInstanceProfile{
 									Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
 								},
-								InstanceId:   aws.String("two"),
-								InstanceType: aws.String("m5.large"),
-								SubnetId:     aws.String("subnet-1"),
-								ImageId:      aws.String("ami-1"),
+								InstanceId:     aws.String("two"),
+								InstanceType:   aws.String("m5.large"),
+								SubnetId:       aws.String("subnet-1"),
+								ImageId:        aws.String("ami-1"),
+								RootDeviceName: aws.String("device-1"),
+								BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+									{
+										DeviceName: aws.String("device-1"),
+										Ebs: &ec2.EbsInstanceBlockDevice{
+											VolumeId: aws.String("volume-1"),
+										},
+									},
+								},
 							},
 						},
 					}, nil)
+
 				m.WaitUntilInstanceRunningWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
 					Return(nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {
@@ -651,15 +768,36 @@ func TestCreateInstance(t *testing.T) {
 								IamInstanceProfile: &ec2.IamInstanceProfile{
 									Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
 								},
-								InstanceId:   aws.String("two"),
-								InstanceType: aws.String("m5.large"),
-								SubnetId:     aws.String("subnet-1"),
-								ImageId:      aws.String("ami-1"),
+								InstanceId:     aws.String("two"),
+								InstanceType:   aws.String("m5.large"),
+								SubnetId:       aws.String("subnet-1"),
+								ImageId:        aws.String("ami-1"),
+								RootDeviceName: aws.String("device-1"),
+								BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+									{
+										DeviceName: aws.String("device-1"),
+										Ebs: &ec2.EbsInstanceBlockDevice{
+											VolumeId: aws.String("volume-1"),
+										},
+									},
+								},
 							},
 						},
 					}, nil)
+
 				m.WaitUntilInstanceRunningWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
 					Return(nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {
@@ -764,15 +902,36 @@ func TestCreateInstance(t *testing.T) {
 								IamInstanceProfile: &ec2.IamInstanceProfile{
 									Arn: aws.String("arn:aws:iam::123456789012:instance-profile/foo"),
 								},
-								InstanceId:   aws.String("two"),
-								InstanceType: aws.String("m5.large"),
-								SubnetId:     aws.String("subnet-1"),
-								ImageId:      aws.String("ami-1"),
+								InstanceId:     aws.String("two"),
+								InstanceType:   aws.String("m5.large"),
+								SubnetId:       aws.String("subnet-1"),
+								ImageId:        aws.String("ami-1"),
+								RootDeviceName: aws.String("device-1"),
+								BlockDeviceMappings: []*ec2.InstanceBlockDeviceMapping{
+									{
+										DeviceName: aws.String("device-1"),
+										Ebs: &ec2.EbsInstanceBlockDevice{
+											VolumeId: aws.String("volume-1"),
+										},
+									},
+								},
 							},
 						},
 					}, nil)
+
 				m.WaitUntilInstanceRunningWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
 					Return(nil)
+
+				m.DescribeVolumes(gomock.Eq(&ec2.DescribeVolumesInput{
+					VolumeIds: []*string{aws.String("volume-1")},
+				})).Return(&ec2.DescribeVolumesOutput{
+					Volumes: []*ec2.Volume{
+						{
+							VolumeId: aws.String("volume-1"),
+							Size:     aws.Int64(60),
+						},
+					},
+				}, nil)
 			},
 			check: func(instance *infrav1.Instance, err error) {
 				if err != nil {

